drivers/epd_bw_spi/epd_bw_spi.c

```diff
@@ -0,0 +1,290 @@
+
+#include "epd_bw_spi.h"
+#include "epd_bw_spi_params.h"
+#include "epd_bw_spi_internal.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void epd_bw_spi_cmd_start(epd_bw_spi_params_t *p, uint8_t cmd, bool cont)
+{
+    DEBUG("[epd_bw_spi] cmd_start: command 0x%02x\n", cmd);
+    if (p->busy_pin != GPIO_UNDEF) {
+        while (gpio_read(p->busy_pin)) {}
+    }
+    gpio_clear(p->dc_pin);
+    spi_transfer_byte(p->spi, p->cs_pin, cont, (uint8_t)cmd);
+    gpio_set(p->dc_pin);
+}
+
+void epd_bw_spi_write_cmd(epd_bw_spi_params_t *p, uint8_t cmd, const uint8_t *params, size_t plen)
+{
+    spi_acquire(p->spi, p->cs_pin, SPI_MODE_0, p->spi_clk);
+    epd_bw_spi_cmd_start(p, cmd, plen > 0);
+    if (plen) {
+        spi_transfer_bytes(p->spi, p->cs_pin, false, params, NULL, plen);
+    }
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_read_cmd(epd_bw_spi_params_t *p, uint8_t cmd, uint8_t *params, size_t plen)
+{
+    spi_acquire(p->spi, p->cs_pin, SPI_MODE_0, p->spi_clk);
+    epd_bw_spi_cmd_start(p, cmd, true);
+    spi_transfer_bytes(p->spi, p->cs_pin, false, NULL, params, plen);
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_wait(epd_bw_spi_params_t *p, uint32_t usec)
+{
+    if (p->busy_pin != GPIO_UNDEF) {
+        DEBUG("[epd_bw_spi] wait: for busy bin\n");
+        while (gpio_read(p->busy_pin)) {}
+    }
+    else {
+        DEBUG("[epd_bw_spi] wait: for %lu microseconds\n", usec);
+        xtimer_usleep(usec);
+    }
+}
+
+int epd_bw_spi_init(epd_bw_spi_t *dev,
+                    const epd_bw_spi_params_t *params,
+                    const epd_bw_spi_controller_t *controller,
+                    uint8_t size_x, uint16_t size_y, epd_bw_spi_entry_mode_t entry_mode)
+{
+    memcpy(&dev->params, params, sizeof(epd_bw_spi_params_t));
+    memcpy(&dev->controller, controller, sizeof(epd_bw_spi_params_t));
+    dev->size_x = size_x;
+    dev->size_y = size_y;
+    dev->entry_mode = entry_mode;
+
+    if (dev->params.rst_pin != GPIO_UNDEF) {
+        if (gpio_init(dev->params.rst_pin, GPIO_OUT) != 0) {
+            DEBUG("[epd_bw_spi] init: error initializing the RST pin\n");
+            return EPD_BW_SPI_RST_FAIL;
+        }
+        gpio_set(dev->params.rst_pin);
+    }
+
+    if (dev->params.busy_pin != GPIO_UNDEF) {
+        if (gpio_init(dev->params.busy_pin, GPIO_IN) != 0) {
+            DEBUG("[epd_bw_spi] init: error initializing the BUSY pin\n");
+            return EPD_BW_SPI_BUSY_FAIL;
+        }
+    }
+
+    if (dev->params.dc_pin == GPIO_UNDEF || gpio_init(dev->params.dc_pin, GPIO_OUT) != 0) {
+        DEBUG("[epd_bw_spi] init: error initializing the DC pin\n");
+        return EPD_BW_SPI_DC_FAIL;
+    }
+    gpio_set(dev->params.dc_pin);
+
+    int res = spi_init_cs(dev->params.spi, dev->params.cs_pin);
+    if (res != SPI_OK) {
+        DEBUG("[epd_bw_spi] init: error initializing the CS pin [%i]\n", res);
+        return res;
+    }
+
+    return 0;
+}
+
+void epd_bw_spi_display_init(epd_bw_spi_t *dev)
```

Comment:
@Citrullin made a good point in the original implementation:
> The factory firmware of good-display pulls the RST pin down and up here. I would use this here too. The wake up function would make sense here then.

The reason I did not do this is that the display does not have to be in sleep here. One display I have does not actually have a reset pin, for example. I see the case for calling `epd_bw_spi_wake` here though, as it doesn't hurt an awake display.